### ISSUE TITLE
Project notebooks stuck "Loading"

### ISSUE THREAD

**Issue by smokedsalmonbagel** [sagemathinc/cocalc-docker#89]
I am trying to debug an issue where every once in a while all files (notebooks, text files etc.) just show a loading screen when opened.  Right before this occurs I see "not saved" in red in the notebook.  It sometimes wont happen for days, other times several times within an hour.  
I tried 
- restarting the project
- refreshing the page
- using a different browser
The only thing that seems to work is waiting (a long time - 30+ minutes?) and restarting the server.
It only affects one project at a time it seems.

Because it happens rarely it is fairly hard to find the issue in the log.  I am going to try to get a log dump next time it happens.  I there another place I should be looking to debug this?  Is there some sort of timeout that can be adjusted to prevent the UI from hanging?

**Comment by williamstein** [comment]
This is https://github.com/sagemathinc/cocalc/issues/4630

Since it isn't a cocalc-specific bug, I'm going to close the issue here.  If you find out anything post it there.  I hope to fix this bug this week; I'm sure I can if I can only reproduce it...

**Comment by williamstein** [comment]
I should re-open this,  since https://github.com/sagemathinc/cocalc/issues/4630 has the same symptom, but it is also resolved by any of of the steps you listed as not working.

**Comment by smokedsalmonbagel** [comment]
I am going to build a new version of cocalc, wait for it to happen again and report back.  Is there a certain commit I should build from to test?

**Comment by williamstein** [comment]
>  Is there a certain commit I should build from to test?

I would suggest building from master as I write this, which is c5174886fb7a2a9f23ce0ab497bed2e18a923953

A few minutes ago I did just push a brand new cocalc-docker to dockerhub, by the way.  

It would be useful to have a copy of your /var/log/hub.log when the problem occurs.

**Comment by smokedsalmonbagel** [comment]
I pulled the latest code and rebuilt cocalc (I am also debugging email invite sending).
Just a moment ago I went to run a Jupyter notebook cell the problem occurred - I could tell because I saw "Not saved" a the top of the notebook.  I already had a terminal open (putty, not within cocalc) and got the exact server time.  Here is the log for the 30 second window in which it happened.  
Symptoms:
- Cannot save or run cells in current notebook
- Cannot view other notebook tabs
- Can see existing files (not sure if this is because they were already loaded?)
- Cannot view static files (tried PNG)
- Terminal 'files' load but are unresponsive
- Restating project does not resolve issue
- Issue affects other projects

Note: I restarted the server, tried going back and running the same cell to see if I would happen again and it did not.

I don't think the log contains anything particularly sensitive so I'll just post it here.
[debuglog.txt](https://github.com/sagemathinc/cocalc-docker/files/4832405/debuglog.txt)

**Comment by williamstein** [comment]
I found an error in the log you posted and pushed [a fix](https://github.com/sagemathinc/cocalc/commit/ee06042293a0b2c2280ae69ea34a1f4872528e65).   It was a mistake in how logging works for the **fix** I wrote for original problem.  I'm building a new cocalc-docker image and intend to push it to Dockerhub today; it might fix your problem.  Definitely the bug in logging would cause this exact problem. 

Rant: the reason this happens is because this code is in CoffeeScript using callbacks instead of Typescript + async/await, and the only way to robustly fix all issues like this is rewriting all of this 30K+ lines of backend code in Typescript using modern async/await, and that's going to take time and more customers to fund it happening.   

We've already rewritten about 90% of the frontend this way (yeah!), so it's very possible.

**Comment by williamstein** [comment]
New version pushed to dockerhub.

**Comment by smokedsalmonbagel** [comment]
Thanks for the fix !  I will grab it and rebuild since I have some small local code changes I would like to include.

> and the only way to robustly fix all issues like this is rewriting all of this 30K+ lines of backend code in Typescript using modern async/await, 
Good to know - I am certainly willing to help out with debugging as much as possible - the project could really be useful for us - and very cool!  Right now we have a few users running non critical scripts - so it is perfect to test.

I will deploy and continue to test for this bug.

**Comment by williamstein** [comment]
>  I have some small local code changes I would like to include.

PR's welcome :-)

>  I am certainly willing to help out with debugging as much as possible

Thanks.  Despite big code quality improvement plans, I'm of course also going to make small fixes to address any bugs that come up.

**Comment by smokedsalmonbagel** [comment]
Just wanted to give an update - we haven't seen this issue since I incorporated [this commit](https://github.com/sagemathinc/cocalc/commit/ee06042293a0b2c2280ae69ea34a1f4872528e65).  
Sometimes there are some odd results with autosaves overwriting other users code but I think this is due to network delays.  I think we are good to close the issue.  Thanks @williamstein !

**Comment by williamstein** [comment]
> Just wanted to give an update - we haven't seen this issue since I incorporated this commit.

Excellent!

> Sometimes there are some odd results with autosaves overwriting other users code

Can you elaborate on this a bit?  CoCalc's realtime sync algorithm is fairly sophisticated (see [unliked Databricks](https://github.com/jupyterlab/rtc/issues/3#issuecomment-665431402)), and so situations where other user code is overwritten should be very rare.

**Comment by smokedsalmonbagel** [comment]
Good to know - I saw it happen a few times.  I would be collaborating live with another user (we were actually in the same room sometimes or on the phone)  I would write a line of code, I would see the NOT saved for just a second and then it would seem to save but when I ran my code I noticed my changes were not there.  
I think based on my experience it seemed like it could have been blamed on dropped messages / spotty network.  Usually it seemed to happen only when the cursors were very close to the same area - it is also possible it was accidental user error.   I will watch more closely for this as we continue to test.
